api/app/controllers/repositories_controller.rb

```diff
@@ -23,4 +23,14 @@ def create
       render(json: { message: "could not create repository.", errors: repositoy.errors.full_messages }, status: 400)
     end
   end
+
+  def sync
+    return render_repository_not_found unless current_repository
+
+    Gitland::Repository.new(current_repository)
+      .then(&:pull)
+      .then { RepositorySyncJob.perform_later(repository_id: current_repository.id) }
```

Comment:
We can't call `#pull` directly in the controller action because `git pull` can take a while to finish.

We should call `#pull` directly inside the `RepositorySyncService` instead. The service should:
1. run `git pull`
2. get the latest known commit hash in our db for the repository
3. run `git log <commit hash>..HEAD`. This will generate the log file for each commit between the last commit we know and the current HEAD.

Reply:
Are you suggesting to do those 3 steps inside those 2 methods?
![image](https://github.com/user-attachments/assets/80f48dd5-fbfb-43e9-92fa-7a80b6d347bc)


btw, those 2 methods seem to have not insignificant code duplication. refactor?

Reply:
>Are you suggesting to do those 3 steps inside those 2 methods?

Almost!

1. We should just call `git pull` once, at the very beginning, so that we update the repository on disk.
2. Inside both extract functions, we should tweak the `git log` call so that it generates the missing logs in our database. Right now by default it generates the full log of the repository. We should just generate the log since the latest commit we know of the repository.

What I have in mind: 
```ruby
def index
  gitland_repository = Gitland::Repository.new(@repository)
  gitland_repository.pull

  extract_full_commit_history(gitland_repository)
  extract_commit_history_for_changes_ledger(gitland_repository)
end
```

```ruby
def extract_full_commit_history(gitland_repository)
  batch = Batch.new
  latest_commit_hash = @repository.commits.last.commit_hash
  current_commit_hash = nil

  gitland_repository.log(from_commit: latest_commit_hash, format: "||%H||%aN||%cs||%as||%P||%s") do |logs|
    # ... rest of the code
  end
```

---

> btw, those 2 methods seem to have not insignificant code duplication. refactor?

If you want to try to refactor this, go for it, but please do it in a separate PR so that we don't mix new features with refactors. Otherwise the diff will become messy.